### DISCUSSION TITLE
Make path method working for StringIO in test memory storage

### DISCUSSION THREAD

**Discussion by dmitry** [shrinerb/shrine#612] [Q&A] [answered]
It is not possible to pass the path to the `send_file` method when using the `memory` storage. Currently, I have the following code:

```ruby
send_file(
  resume.file.to_io.path
)
```

I want to use the `x-accel-redirect` header feature in Nginx to pass the `io.path` location to Nginx and let it stream the data, rather than streaming the data directly from the Rails application.

In all environments except for `test`, the file system is used. The `test` environment uses the memory storage, which returns a `StringIO` object, which does not have a file associated with it and therefore does not have a `path`. What is the best solution to make this work? Is using `is_a?(StringIO)` condition the only option, or is there something better?

**Comment by janko** [comment, marked as answer]
Yeah, the memory storage cannot retrieve the path, because the uploaded file doesn't exist on the filesystem. You'll need to create a conditional for memory storage, something like:

```rb
class ApplicationController < ActionController::Base
  private

  def send_upload(uploaded_file, ...)
    if uploaded_file.storage.is_a?(Shrine::Storage::Memory)
      send_file uploaded_file.download.path, ...
    else
      send_file uploaded_file.to_io.path, ...
    end
  end
end
```
```rb
# in your controller:
send_upload(resume.file)
```

You could also use `send_data uploaded_file.read, ...` for the memory storage path.

**Comment by dmitry** [comment]
@janko thank you for the quick response, seems like that's the only way to do it.

That's what I did:

```ruby
    send_upload(
      file,
      type: inline ? 'application/pdf' : 'application/force-download',
      disposition: inline ? :inline : :attachment,
      filename: "#{sanitize_filename(filename)}.pdf"
    )
  end

  private

  def send_upload(uploaded_file, **options)
    if uploaded_file.storage.is_a?(Shrine::Storage::Memory)
      send_file uploaded_file.download.path, **options
    else
      send_file uploaded_file.to_io.path, **options
    end
  end
```
